lib/backends/webgl/ops/unpack.ts

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+import {Tensor} from '../../../tensor';
+import {WebGLInferenceHandler} from '../inference-handler';
+import {ProgramInfo, RunData, WebGLOperator} from '../types';
+import {getCoordsDataType} from '../utils';
+
+import {getChannels} from './packing_utils';
+
+export class WebGLUnpack implements WebGLOperator {
+  run(inferenceHandler: WebGLInferenceHandler, inputs: Tensor[]): Tensor[] {
+    return inferenceHandler.run(this, inputs);
+  }
+  createProgramInfo(handler: WebGLInferenceHandler, inputs: Tensor[]): ProgramInfo {
+    if (inputs.length !== 1) {
+      throw new Error(`Pack kernel should have input tensor count to 1.`);
+    }
+
+    const inputTexture = handler.getTextureData(inputs[0].dataId);
+    if (!inputTexture) {
+      throw new Error(`packed input texture must exist`);
+    }
+
+    const outputLayout = handler.createTextureLayoutFromShape(inputTexture.unpackedShape);
+    const outputShape = outputLayout.shape;
+    const rank = outputShape.length;
+
+    const channels = getChannels('rc', rank);
+    const innerDims = channels.slice(-2);
+    const coordsDataType = getCoordsDataType(rank);
+    const unpackChannel = unpackFromChannel(rank);
+    const sourceCoords = getSourceCoords(rank, channels);
+    const coords = rank <= 1 ? 'rc' : `vec2(${innerDims.join(',')})`;
+    const shaderSource = `
+        ${unpackChannel}
+        void main() {
+          ${coordsDataType} rc = getOutputCoords();
+
+          // Sample the texture with the coords to get the rgba channel value.
+          vec4 packedInput = getA(${sourceCoords});
+
+          outputColor = vec4(getChannel(packedInput, ${coords}), 0, 0, 0);
+        }
+      `;
+
+    return {
+      inputLayouts: [handler.getOrCreateTextureLayout(inputs[0])],
+      outputLayout,
+      samplers: ['A'],
+      shaderSource,
+      hasMain: true,
+      isInputsPacked: true,
+      isOutputPacked: false,
+    };
+  }
+  createRunData(handler: WebGLInferenceHandler, programInfo: ProgramInfo, inputs: Tensor[]): RunData {
+    const inputTDs = [handler.getOrCreateTextureData(inputs[0], programInfo.inputLayouts[0])];
+    return {
+      inputTextureDatas: inputTDs,
+      outputTextureData: handler.createTextureDataFromLayout(programInfo.outputLayout, inputTDs[0].tensor.type),
+      uniformData: {}
+    };
+  }
+}
+
+function unpackFromChannel(rank: number): string {
+  if (rank <= 1) {
+    return `
+    float getChannel(vec4 frag, int dim) {
+      float modCoord = mod(float(dim), 2.);
+      return modCoord == 0. ? frag.r : frag.g;
+    }
+  `;
+  }
+  return `
+  float getChannel(vec4 frag, vec2 innerDims) {
+    vec2 modCoord = mod(innerDims, 2.);
```

Comment:
>mod [](start = 20, length = 3)

[yulong] double check if mod support int value.

Reply:
mod only supports vector of floating point operations. imod only support single integer operations. So I'll keep innerDims of type vec2 and leave mod as it is here.

---
In reply to: [586931791](https://github.com/microsoft/onnxjs/pull/266#discussion_r586931791) [](ancestors = 586931791)